src/passes/JSPI.cpp

```diff
@@ -0,0 +1,171 @@
+// `WebAssembly.Function`. All imports will also be wrapped, but they will take
+// the stored suspender and pass it as the first param to the imported function
+// that should be created from a "suspending" `WebAssembly.Function`.
+//
+namespace wasm {
+
+struct JSPI : public Pass {
+
+  JSPI() {}
+
+  Type externref = Type(HeapType::ext, Nullable);
+
+  void run(PassRunner* runner, Module* module) override {
+    Builder builder(*module);
+    // Create a global to store the suspender that is passed into exported
+    // functions and will then need to be passed out to the imported functions.
+    Name suspender = Names::getValidGlobalName(*module, "suspender");
+    module->addGlobal(builder.makeGlobal(
+      suspender, externref, builder.makeRefNull(externref), Builder::Mutable));
+
+    // Wrap each exported function in a function that stores the suspender
+    // and calls the original export.
+    std::vector<std::unique_ptr<Export>> newExports;
+    for (auto& ex : module->exports) {
+      if (ex->kind == ExternalKind::Function) {
+        auto* func = module->getFunction(ex->value);
+        auto legalName = makeWrapperForExport(func, module, suspender);
+        ex->value = legalName;
+      }
+    }
+    for (auto& ex : newExports) {
+      module->addExport(std::move(ex));
+    }
+
+    // Avoid iterator invalidation later.
+    std::vector<Function*> originalFunctions;
+    for (auto& func : module->functions) {
+      originalFunctions.push_back(func.get());
+    }
+    // Wrap each imported function in a function that gets the global suspender
+    // and passes it on to the imported function.
+    for (auto* im : originalFunctions) {
+      if (im->imported()) {
+        auto funcName = makeWrapperForImport(im, module, suspender);
+        ElementUtils::iterAllElementFunctionNames(module, [&](Name& name) {
+          if (name == im->name) {
+            name = funcName;
+          }
+        });
+      }
+    }
+  }
+
+private:
+  Name makeWrapperForExport(Function* func, Module* module, Name suspender) {
+    Name wrapperName = Names::getValidFunctionName(
+      *module, std::string("export$") + func->name.str);
+
+    // a method may be exported multiple times
+    if (module->getFunctionOrNull(wrapperName)) {
+      return wrapperName;
+    }
+
+    Builder builder(*module);
+    auto* wrapperFunc = new Function();
+    wrapperFunc->name = wrapperName;
+
+    auto* call = module->allocator.alloc<Call>();
+    call->target = func->name;
+    call->type = func->getResults();
+
+    // Add an externref param as the first argument and copy all the original
+    // params to new export.
+    std::vector<Type> wrapperParams;
+    wrapperParams.push_back(externref);
+    Index i = 1;
+    for (const auto& param : func->getParams()) {
+      call->operands.push_back(
+        builder.makeLocalGet(wrapperParams.size(), param));
+      wrapperParams.push_back(param);
```

Comment:
It might be nice to copy the param names to the wrapper as well to aid in debugging.